docs/proposals/20201024-advancedcronjob.md

```diff
@@ -0,0 +1,178 @@
+
+## Motivation
+
+- Developer may come across a use-case when some job needs to be executed on at a specific schedule.
+- Found same use-case requirement in Issues #251 and I got motivated to implement it
+
+### Goals
+
+- Implementing a custom controller for AdvancedCronJob which acts like CronJob but it schedules Job/BroadcastJob or other CRD
+
+## Proposal
+
+- Adding a new CRD and controller for AdvancedCronJob.
+- AdvancedCronJob should be able to reconcile Job/BroadcastJob or any other future CRD if required.
+- Once AdvancedCronJob is created, spec cannot be modified.
+- Adding webhook for validation of AdvancedCronJob.
+
+### User Stories
+
+- Implement a CRD which contains below fields and a controller which honors all the fields and reconciles accordingly.
+```
+
+// AdvancedCronJob is the Schema for the advancedcronjobs API
+type AdvancedCronJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AdvancedCronJobSpec   `json:"spec,omitempty"`
+	Status AdvancedCronJobStatus `json:"status,omitempty"`
+}
+
+
+type AdvancedCronJobSpec struct {
+	Schedule string `json:"schedule" protobuf:"bytes,1,opt,name=schedule"`
+
+	// Optional deadline in seconds for starting the job if it misses scheduled
+	// time for any reason.  Missed jobs executions will be counted as failed ones.
+	// +optional
+	StartingDeadlineSeconds *int64 `json:"startingDeadlineSeconds,omitempty" protobuf:"varint,2,opt,name=startingDeadlineSeconds"`
+
+	// Specifies how to treat concurrent executions of a Job.
+	// Valid values are:
+	// - "Allow" (default): allows CronJobs to run concurrently;
+	// - "Forbid": forbids concurrent runs, skipping next run if previous run hasn't finished yet;
+	// - "Replace": cancels currently running job and replaces it with a new one
+	// +optional
+	ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy,omitempty" protobuf:"bytes,3,opt,name=concurrencyPolicy"`
+
+	// Paused will pause the cron job.
+	// +optional
+	Paused bool `json:"paused,omitempty" protobuf:"bytes,4,opt,name=paused"`
+
+	// +kubebuilder:validation:Minimum=0
+
+	// The number of successful finished jobs to retain.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	SuccessfulJobsHistoryLimit *int32 `json:"successfulJobsHistoryLimit,omitempty" protobuf:"varint,5,opt,name=successfulJobsHistoryLimit"`
+
+	// +kubebuilder:validation:Minimum=0
+
+	// The number of failed finished jobs to retain.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	// +optional
+	FailedJobsHistoryLimit *int32 `json:"failedJobsHistoryLimit,omitempty" protobuf:"varint,6,opt,name=failedJobsHistoryLimit"`
+
+	// Specifies the job that will be created when executing a CronJob.
+	Template AdvancedCronJobTemplate `json:"template" protobuf:"bytes,7,opt,name=jobTemplate"`
+}
+
+type AdvancedCronJobTemplate struct {
```

Comment:
How about AdvancedCronJobTemplate -> CronJobTemplate?

Reply:
Done in commit https://github.com/openkruise/kruise/pull/437/commits/5b0344eca29b2fc30c6a3f70846dc911779f6e3e